src/realm/db.cpp

```diff
@@ -2332,13 +2333,28 @@ TransactionRef DB::start_frozen(VersionID version_id)
 {
     if (!is_attached())
         throw LogicError(LogicError::wrong_transact_state);
+    // prune stale frozen transactions
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
+    while (!m_frozen_transactions.empty() && m_frozen_transactions.begin()->second.expired()) {
+        m_frozen_transactions.erase(m_frozen_transactions.begin());
+    }
     ReadLockInfo read_lock;
-    grab_read_lock(read_lock, version_id);
     ReadLockGuard g(*this, read_lock);
+    grab_read_lock(read_lock, version_id);
+    version_id.version = read_lock.m_version;
+    // reuse existing frozen transaction if one exists:
+    std::weak_ptr<Transaction>& weak_tr = m_frozen_transactions[version_id.version];
+    TransactionRef res = weak_tr.lock();
+    if (res) {
+        return res; // <-- also releases read_lock - the now shared frozen transaction has its own
+    }
+    // none cached, create a new. In case we haven't grabbed the read_lock yet, we do that first.
     Transaction* tr = new Transaction(shared_from_this(), &m_alloc, read_lock, DB::transact_Frozen);
     tr->set_file_format_version(get_file_format_version());
+    res = TransactionRef(tr, TransactionDeleter);
```

Comment:
Can this be merged with line 2352 like so:

```cpp
auto res = TransactionRef(new Transaction(shared_from_this(), &m_alloc, read_lock, DB::transact_Frozen), TransactionDeleter);
res->set_file_format_version(get_file_format_version());
```

Then there would be no period where `tr` wouldn't be owned by a smart pointer. Right now if line 2353 threw, there would be a leak. This may be impossible, but it requires being very careful to never add anything that throws in before it gets managed.